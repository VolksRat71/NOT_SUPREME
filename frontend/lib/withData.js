import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { endpoint } from '../config';
import { LOCAL_STATE_QUERY, TOGGLE_CART_MUTATION } from '../components/Cart';

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    // Local data
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_, variables, { cache }) {
            //read cart open value from cache
            const { cartOpen } = cache.readQuery({
              query: LOCAL_STATE_QUERY,
            });
            const data = {
              data: { cartOpen: !cartOpen }
            };
            cache.writeData(data);
            return data;
          }
        }
      },
      defaults: {
        cartOpen: false,
      }
    }
  });
}

export default withApollo(createClient);