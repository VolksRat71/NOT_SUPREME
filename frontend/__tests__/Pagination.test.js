import { mount } from 'enzyme';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import toJSON from 'enzyme-to-json';
import Router from 'next/router'
import { fakeUser } from '../lib/testUtils';
import Pagination, { PAGINATION_QUERY } from '../components/Pagination';

// * Overiding Router prototypes for tests
Router.router = {
    push() { },
    prefetch() { }
}

function makeMocksFor(length) {
    return [
        {
            request: { query: PAGINATION_QUERY },
            result: {
                data: {
                    itemsConnection: {
                        __typename: 'aggregate',
                        aggregate: {
                            count: length,
                            __typename: 'count',
                        },
                    },
                },
            },
        },
    ];
}

describe('<Pagination /> ', () => {
    test('Displays a loading message ', () => {
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(1)}>
                <Pagination />
            </MockedProvider>
        )
        expect(wrapper.text()).toContain('Loading...')
    })

    test('Renders Pagination for 18 items  ', async () => {
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(18)}>
                <Pagination page={1} />
            </MockedProvider>
        )
        await wait()
        wrapper.update()
        expect(wrapper.find('.totalPages').text()).toEqual('5');
        const pagination = wrapper.find('[data-test="pagination"]');
        expect(toJSON(pagination)).toMatchSnapshot()
    });

    test('disables prev button on first page', async () => {
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(18)}>
                <Pagination page={1} />
            </MockedProvider>
        );
        await wait();
        wrapper.update();
        expect(wrapper.find('a.prev').prop('aria-disabled')).toEqual(true);
        expect(wrapper.find('a.next').prop('aria-disabled')).toEqual(false);
    });

    test('disables next button on last page', async () => {
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(18)}>
                <Pagination page={5} />
            </MockedProvider>
        );
        await wait();
        wrapper.update();
        expect(wrapper.find('a.prev').prop('aria-disabled')).toEqual(false);
        expect(wrapper.find('a.next').prop('aria-disabled')).toEqual(true);
    });

    test('enables all buttons on a middle page', async () => {
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(18)}>
                <Pagination page={3} />
            </MockedProvider>
        );
        await wait();
        wrapper.update();
        expect(wrapper.find('a.prev').prop('aria-disabled')).toEqual(false);
        expect(wrapper.find('a.next').prop('aria-disabled')).toEqual(false);
    });
})
