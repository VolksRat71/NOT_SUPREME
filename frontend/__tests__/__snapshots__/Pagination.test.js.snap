// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination />  Renders Pagination for 18 items   1`] = `
Array [
  <PaginationStyles
    data-test="pagination"
  >
    <div
      className="PaginationStyles-sc-1pn2n7t-0 kvIfKH"
      data-test="pagination"
    >
      <SideEffect(Head)>
        <Head />
      </SideEffect(Head)>
      <Link
        href={
          Object {
            "pathname": "items",
            "query": Object {
              "page": 0,
            },
          }
        }
        prefetch={true}
      >
        <a
          aria-disabled={true}
          className="prev"
          href="items?page=0"
          onClick={[Function]}
        >
          ← Prev
        </a>
      </Link>
      <p>
        Page 
        1
         of 
        <span
          className="totalPages"
        >
          5
        </span>
      </p>
      <p>
        18
         Items Total
      </p>
      <Link
        href={
          Object {
            "pathname": "items",
            "query": Object {
              "page": 2,
            },
          }
        }
        prefetch={true}
      >
        <a
          aria-disabled={false}
          className="next"
          href="items?page=2"
          onClick={[Function]}
        >
          Next →
        </a>
      </Link>
    </div>
  </PaginationStyles>,
  <div
    className="PaginationStyles-sc-1pn2n7t-0 kvIfKH"
    data-test="pagination"
  >
    <SideEffect(Head)>
      <Head />
    </SideEffect(Head)>
    <Link
      href={
        Object {
          "pathname": "items",
          "query": Object {
            "page": 0,
          },
        }
      }
      prefetch={true}
    >
      <a
        aria-disabled={true}
        className="prev"
        href="items?page=0"
        onClick={[Function]}
      >
        ← Prev
      </a>
    </Link>
    <p>
      Page 
      1
       of 
      <span
        className="totalPages"
      >
        5
      </span>
    </p>
    <p>
      18
       Items Total
    </p>
    <Link
      href={
        Object {
          "pathname": "items",
          "query": Object {
            "page": 2,
          },
        }
      }
      prefetch={true}
    >
      <a
        aria-disabled={false}
        className="next"
        href="items?page=2"
        onClick={[Function]}
      >
        Next →
      </a>
    </Link>
  </div>,
]
`;
